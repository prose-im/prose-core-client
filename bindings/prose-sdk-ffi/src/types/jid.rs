// prose-core-client/prose-sdk-ffi
//
// Copyright: 2023, Marc Bauer <mb@nesium.com>
// License: Mozilla Public License v2.0 (MPL v2.0)

use jid::{BareJid, DomainPart, Error as JidParseError, NodePart};

#[derive(Debug, Clone, PartialEq)]
pub struct JID {
    pub node: Option<String>,
    pub domain: String,
}

impl JID {
    pub fn to_bare(&self) -> Result<BareJid, JidParseError> {
        Ok(BareJid::from_parts(
            self.node
                .as_ref()
                .map(|node| NodePart::new(node))
                .transpose()?
                .as_ref(),
            &DomainPart::new(&self.domain)?,
        ))
    }
}

#[uniffi::export]
pub fn parse_jid(jid: String) -> Result<JID, JidParseError> {
    Ok(jid.parse::<BareJid>()?.into())
}

impl From<BareJid> for JID {
    fn from(value: BareJid) -> Self {
        JID {
            node: value.node().map(|s| s.to_string()),
            domain: value.domain().to_string(),
        }
    }
}

impl From<JID> for BareJid {
    fn from(value: JID) -> Self {
        BareJid::from_parts(
            value
                .node
                .as_ref()
                .map(|node| NodePart::new(node).unwrap())
                .as_ref(),
            &DomainPart::new(&value.domain).unwrap(),
        )
    }
}

impl ToString for JID {
    fn to_string(&self) -> String {
        BareJid::from(self.clone()).into_inner()
    }
}

#[uniffi::export]
pub fn format_jid(jid: JID) -> String {
    jid.to_string()
}
