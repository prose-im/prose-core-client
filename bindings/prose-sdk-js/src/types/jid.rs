// prose-core-client/prose-sdk-js
//
// Copyright: 2023, Marc Bauer <mb@nesium.com>
// License: Mozilla Public License v2.0 (MPL v2.0)

use core::fmt::{Debug, Display, Formatter};
use core::str::FromStr;

use wasm_bindgen::prelude::*;

use prose_core_client::dtos::{RoomId, UserId};

#[derive(Debug, PartialEq, Clone)]
#[wasm_bindgen(js_name = "JID")]
pub struct BareJid(jid::BareJid);

#[wasm_bindgen(js_class = "JID")]
impl BareJid {
    #[wasm_bindgen(constructor)]
    pub fn new(str: &str) -> Result<BareJid, JsError> {
        Ok(Self(jid::BareJid::from_str(str)?))
    }

    /// The node part of the Jabber ID, if it exists, else None.
    #[wasm_bindgen(getter)]
    pub fn node(&self) -> Option<String> {
        self.0.node_str().map(ToString::to_string)
    }

    /// The domain of the Jabber ID.
    #[wasm_bindgen(getter)]
    pub fn domain(&self) -> String {
        self.0.domain_str().to_string()
    }

    #[wasm_bindgen(js_name = "toString")]
    pub fn to_string(&self) -> String {
        self.0.to_string()
    }

    pub fn equals(&self, other: &BareJid) -> bool {
        self == other
    }
}

impl BareJid {
    pub fn to_full_jid_with_resource(&self, resource: &jid::ResourcePart) -> jid::FullJid {
        jid::FullJid::from_parts(self.0.node().as_ref(), &self.0.domain(), resource)
    }
}

impl From<jid::BareJid> for BareJid {
    fn from(value: jid::BareJid) -> Self {
        Self(value)
    }
}

impl From<&jid::BareJid> for BareJid {
    fn from(value: &jid::BareJid) -> Self {
        Self(value.clone())
    }
}

impl From<&BareJid> for jid::BareJid {
    fn from(value: &BareJid) -> Self {
        value.0.clone()
    }
}

impl From<BareJid> for jid::BareJid {
    fn from(value: BareJid) -> Self {
        value.0
    }
}

impl From<&BareJid> for jid::Jid {
    fn from(value: &BareJid) -> Self {
        jid::Jid::Bare(value.0.clone())
    }
}

impl AsRef<jid::BareJid> for BareJid {
    fn as_ref(&self) -> &jid::BareJid {
        &self.0
    }
}

impl Display for BareJid {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        Display::fmt(&self.0, f)
    }
}

impl From<BareJid> for RoomId {
    fn from(value: BareJid) -> Self {
        RoomId::from(value.0)
    }
}

impl From<BareJid> for UserId {
    fn from(value: BareJid) -> Self {
        UserId::from(value.0)
    }
}

impl From<&BareJid> for UserId {
    fn from(value: &BareJid) -> Self {
        UserId::from(value.0.clone())
    }
}
