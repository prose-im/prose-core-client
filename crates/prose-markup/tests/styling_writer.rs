// prose-core-client/prose-markup
//
// Copyright: 2023, Marc Bauer <mb@nesium.com>
// License: Mozilla Public License v2.0 (MPL v2.0)

use insta::assert_snapshot;
use jid::BareJid;
use pretty_assertions::assert_eq;

use prose_markup::MarkdownParser;

#[test]
fn test_nested_spans() {
    let parser = MarkdownParser::new(include_str!("fixtures/nested_spans.md"));
    assert_snapshot!(parser.convert_to_message_styling());
    assert!(parser.collect_mentions().is_empty());
}

#[test]
fn test_links() {
    let parser = MarkdownParser::new(include_str!("fixtures/links.md"));
    assert_snapshot!(parser.convert_to_message_styling());
    assert_eq!(
        vec!["user@prose.org".parse::<BareJid>().unwrap()],
        parser.collect_mentions()
    );
}

#[test]
fn test_complex() {
    let parser = MarkdownParser::new(include_str!("fixtures/complex.md"));
    assert_snapshot!(parser.convert_to_message_styling());
    assert!(parser.collect_mentions().is_empty());
}

#[test]
fn test_nested_lists() {
    let parser = MarkdownParser::new(include_str!("fixtures/nested_lists.md"));
    assert_snapshot!(parser.convert_to_message_styling());
    assert!(parser.collect_mentions().is_empty());
}

#[test]
fn test_nested_blockquotes() {
    let parser = MarkdownParser::new(include_str!("fixtures/nested_blockquotes.md"));
    assert_snapshot!(parser.convert_to_message_styling());
    assert!(parser.collect_mentions().is_empty());
}

#[test]
fn test_escapes_html_tags() {
    let parser = MarkdownParser::new(include_str!("fixtures/html_tags.md"));
    assert_snapshot!(parser.convert_to_message_styling());
    assert!(parser.collect_mentions().is_empty());
}

#[test]
fn test_plain_text() {
    let parser = MarkdownParser::new("Hello World");
    assert_eq!("Hello World", parser.convert_to_message_styling());
    assert_eq!("<p>Hello World</p>", parser.convert_to_html());
}

#[test]
fn test_fenced_url_is_not_converted_to_anchor() {
    let parser = MarkdownParser::new("`https://www.example.com`");
    assert_eq!(
        "`https://www.example.com`",
        parser.convert_to_message_styling()
    );
    assert_eq!(
        "<p><code>https://www.example.com</code></p>",
        parser.convert_to_html()
    );
}

#[test]
fn test_encodable_entities() {
    let parser = MarkdownParser::new(
        "Already did, it's done by switching the type attribute between text <> password",
    );
    assert_eq!(
        "Already did, it's done by switching the type attribute between text <> password",
        parser.convert_to_message_styling()
    );
    assert_eq!(
        "<p>Already did, it's done by switching the type attribute between text &lt;&gt; password</p>",
        parser.convert_to_html()
    )
}
