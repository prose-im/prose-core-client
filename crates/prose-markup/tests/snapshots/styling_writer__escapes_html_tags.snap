---
source: crates/prose-markup/tests/styling_writer.rs
expression: parser.convert_to_message_styling()
---
HTML Tags

<table>
    <tr>
        <td><pre>**Hello**, _world_</pre></td>
    </tr>
</table>
